pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/readers/PinotSegmentColumnReader.java

```diff
@@ -95,26 +99,92 @@ public Object getValue(int docId) {
         return values;
       }
     } else {
-      // NOTE: Only support single-value raw index
-      assert _forwardIndexReader.isSingleValue();
-
-      switch (_forwardIndexReader.getStoredType()) {
-        case INT:
-          return _forwardIndexReader.getInt(docId, _forwardIndexReaderContext);
-        case LONG:
-          return _forwardIndexReader.getLong(docId, _forwardIndexReaderContext);
-        case FLOAT:
-          return _forwardIndexReader.getFloat(docId, _forwardIndexReaderContext);
-        case DOUBLE:
-          return _forwardIndexReader.getDouble(docId, _forwardIndexReaderContext);
-        case BIG_DECIMAL:
-          return _forwardIndexReader.getBigDecimal(docId, _forwardIndexReaderContext);
-        case STRING:
-          return _forwardIndexReader.getString(docId, _forwardIndexReaderContext);
-        case BYTES:
-          return _forwardIndexReader.getBytes(docId, _forwardIndexReaderContext);
-        default:
-          throw new IllegalStateException();
+      if (_forwardIndexReader.isSingleValue()) {
+        switch (_forwardIndexReader.getStoredType()) {
+          case INT:
+            return _forwardIndexReader.getInt(docId, _forwardIndexReaderContext);
+          case LONG:
+            return _forwardIndexReader.getLong(docId, _forwardIndexReaderContext);
+          case FLOAT:
+            return _forwardIndexReader.getFloat(docId, _forwardIndexReaderContext);
+          case DOUBLE:
+            return _forwardIndexReader.getDouble(docId, _forwardIndexReaderContext);
+          case BIG_DECIMAL:
+            return _forwardIndexReader.getBigDecimal(docId, _forwardIndexReaderContext);
+          case STRING:
+            return _forwardIndexReader.getString(docId, _forwardIndexReaderContext);
+          case BYTES:
+            return _forwardIndexReader.getBytes(docId, _forwardIndexReaderContext);
+          default:
+            throw new IllegalStateException();
+        }
+      } else {
+        Preconditions.checkState(_maxNumValuesPerMVEntry >= 0, "maxNumValuesPerMVEntry is negative for an MV column.");
+
+        switch (_forwardIndexReader.getStoredType()) {
```

Comment:
Thanks for adding this. I do see value in keeping this code even if you decide to either add more efficient functions to directly fetch the type array or if you decide to directly use the `ForwardIndexReader`.